server/src/main/java/org/elasticsearch/cluster/routing/RoutingNodes.java

```diff
@@ -601,32 +595,35 @@ assert getByAllocationId(failedShard.shardId(), failedShard.allocationId().getId
                 cancelRelocation(sourceShard);
                 remove(failedShard);
             }
-            routingChangesObserver.shardFailed(failedShard, unassignedInfo);
         } else {
             assert failedShard.active();
             if (failedShard.primary()) {
                 // promote active replica to primary if active replica exists
-                ShardRouting activeReplica = activeReplicaWithHighestVersion(failedShard.shardId());
-                if (activeReplica == null) {
-                    moveToUnassigned(failedShard, unassignedInfo);
-                } else {
-                    movePrimaryToUnassignedAndDemoteToReplica(failedShard, unassignedInfo);
-                    promoteReplicaToPrimary(activeReplica, routingChangesObserver);
-                }
+                tryPromoteActiveReplica(failedShard, unassignedInfo, routingChangesObserver);
             } else {
-                assert failedShard.primary() == false;
                 if (failedShard.relocating()) {
                     remove(failedShard);
                 } else {
                     moveToUnassigned(failedShard, unassignedInfo);
                 }
             }
-            routingChangesObserver.shardFailed(failedShard, unassignedInfo);
         }
+        routingChangesObserver.shardFailed(failedShard, unassignedInfo);
         assert node(failedShard.currentNodeId()).getByShardId(failedShard.shardId()) == null : "failedShard " + failedShard +
             " was matched but wasn't removed";
     }
 
+    private void tryPromoteActiveReplica(ShardRouting failedShard, UnassignedInfo unassignedInfo,
```

Comment:
perhaps call this `unassignPrimaryAndPromoteActiveReplicaIfExists`? Also add `assert failedShard.primary()`